### PR TITLE
Scrape check report for outputs from Distributor

### PR DESCRIPTION
**Pull request by benreynolds-drizly**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+
+	// search for the specific installation we care about
+	// spew.Dump(allInstallations)
+
+	// construct client with the installation
+	ntr := ghinstallation.NewFromAppsTransport(itr, *allInstallations[0].ID)
```
Comment (ikatz-drizly) [inline client.go:42]: I see several things going on here that are similar to what I have on the ruby side, so let's compare notes.  I see that you are fetching _all_ pages of installations, but taking just the first installation ID.  Can you find docs to indicate that the first item in the array is always the correct one?  I ended up going a step further and searching for an installation in which the owner was equal to the repo owner.  

TL;DR download all & search for a match, or download just the first page and pick the first element _and_ add a comment about why that's safe

File: main.go [hunk 2]
```diff
@@ -5,93 +5,131 @@ import (
...
 		githubactions.Fatalf("Error disptaching event: %v", err.Error())
 	}
+}
 
-	if inputs.waitForCheck {
-		// Wait for the check to be completed after all actions have passed
-		checkTimeoutCtx, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(inputs.waitTimeoutSeconds))
-		// cancel context just in case it sticks around
-		defer cancel()
+func waitForCheck(client *github.Client, githubVars githubVars, inputs inputs, checkRun *github.CheckRun) {
```
Comment (ikatz-drizly) [inline main.go:99]: We should add some console logging that shows some activity and/or indicates how much time is remaining, maybe every 30-60 seconds.

File: main.go [hunk 3]
```diff
@@ -5,93 +5,131 @@ import (
...
+}
+
+func scrapeOutputs(client *github.Client, githubVars githubVars, checkId int64) {
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	check, _, err := client.Checks.GetCheckRun(ctx, githubVars.repositoryOwner, githubVars.repositoryName, checkId)
+	if err != nil {
+		githubactions.Fatalf("Error fetching check for output scraping: %v", err.Error())
```
Comment (ikatz-drizly) [inline main.go:128]: Since we will be scraping outputs multiple times, should we ignore any errors on individual scrapes?  It seems like this should only catch instances where the check run is accidentally deleted (since upon creation we get its ID)... and without the check run there is no way to signal the original job.  So it seems like we should either make even more noise in this case, or just ignore it until a few more attempts fail.

File: output.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package main
+
+import (
+	"strings"
+)
+
+func parseOutputs(reportText *string) string {
+
+	splitReport := strings.Split(*reportText, "---- BEGIN CHECK OUTPUT ----")
```
Comment (ikatz-drizly) [inline output.go:9]: `BEGIN CHECK OUTPUT` feels like it should (1) be calculated in a single place in code, and (2) include something from an env var as a way of double/triple checking that we are reading and writing the checks that we _expect_ to be reading and writing

**Comment by ikatz-drizly** [comment]
As a meta-comment, I observe that none of the functions have comments describing their purpose nor what the arguments do nor what the return value is.

File: main.go (hunk 2 above)
Comment (benreynolds-drizly) [reply main.go:99]: It already does log progress. I will add time remaining to that:

Screen Shot 2021-08-12 at 8 38 03 AM

File: main.go (hunk 3 above)
Comment (benreynolds-drizly) [reply main.go:128]: It doesn't scrape output multiple times and this is the original job. This is the original job scraping the output after it has already been signaled by distributor that the check is complete. As far as making more noise, this is about as much noise as you can make. This will fail the job in error with the given message (which appears in bright red and very clear in the logs).

File: output.go (hunk 4 above)
Comment (benreynolds-drizly) [reply output.go:9]: I don't understand this comment. The `BEGIN CHECK OUTPUT` is not calculated...?...it's just a constant that @tatchley-drizly and I agreed on using for signaling the start of the outputs section in the report. An env var would not help verify anything about the check. We have the check ID since this is the action that created the check in the first place.

**Comment by benreynolds-drizly** [comment]
> As a meta-comment, I observe that none of the functions have comments describing their purpose nor what the arguments do nor what the return value is.

I'm pretty opinionated about what makes for good, clean code and have founded my opinions based on [Clean Code: A Handbook of Agile Software Craftsmanship](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882) by Uncle Bob. Two of the key points about comments made in this book are "Always try to explain yourself in code" and "Don't add obvious noise" [[reference](https://gist.github.com/wojteklu/73c6914cc446146b8b533c0988cf8d29#comments-rules)]. Comments of the type that I think you are talking about here seem to me to be largely at odds with both of these principles. The method names provide the purpose of the method and listing the arguments and return value(s) wouldn't tell you anything that reading the code couldn't already tell you. If we were building an external library which would be consumed by many customers needing a well documented API then that would be different, but in this case I firmly believe that the code is its own best documentation.

File: main.go (hunk 2 above)
Comment (benreynolds-drizly) [reply main.go:99]: Updated:

Screen Shot 2021-08-12 at 9 34 54 AM

File: client.go (hunk 1 above)
Comment (benreynolds-drizly) [reply client.go]: As far as I can tell there's no reliable way to determine which installation is the correct one. I updated this to only pull the first installation and clarified with additional comments both in the code and in the README. Just for reference, [here is a public action on github marketplace](https://github.com/getsentry/action-github-app-token/blob/main/src/main.ts#L22) doing the same thing.

File: main.go (hunk 2 above)
Comment (ikatz-drizly) [reply main.go:99]: Nice!

File: main.go (hunk 3 above)
Comment (ikatz-drizly) [reply main.go:128]: OK.  I think I'd missed something with regard to fetching the check run status (which happens multiple times) vs getting the contents of the run itself.  For some reason I thought those were the same call

File: output.go (hunk 4 above)
Comment (ikatz-drizly) [reply output.go:9]: Sure... let me clarify.  The first part of my comment is just observing that you define the same string in 2 places and should DRY that up.  

The second part was my general nervousness around using arbitrary strings, because all instances _everywhere_ will have this same string (I've been burned before by a form of false-positive where the right string was found in the wrong place).   In this case, you're right that you need the check ID to even get this string so it shouldn't be a problem. 

But you should still assign `"---- BEGIN CHECK OUTPUT ----"` to a constant in your code, since the whole thing breaks if those 2 strings ever fall out of agreement.

File: client.go (hunk 1 above)
Comment (mikekantzer-drizly) [reply client.go]: the installation is per-org, right? are we expecting it to change? can we hard-code it?

File: wait.go [hunk 5]
```diff
@@ -9,39 +9,41 @@ import (
...
+			return *check.Conclusion == "success", nil
 		}
 
 		// check if context has been closed(either by timeout or another error in err group).
 		// If so, exit with error.
 		// If not, sleep for a bit and loop again
 		select {
 		case <-ctx.Done():
-			return fmt.Errorf("stopping check for id %v: %w", id, ctx.Err())
+			return false, fmt.Errorf("Timeout: %w", ctx.Err())
```
Comment (mikekantzer-drizly) [inline wait.go:43]: `timeout` isn't needed in the error message here. `ctx.Err()` will include information on why `Done()` closed. In the case of a context deadline, it'll say something like "context deadline exceeded".

**Review by mikekantzer-drizly — APPROVED** [review]
Overall, this looks good to me

File: client.go (hunk 1 above)
Comment (benreynolds-drizly) [reply client.go]: You can install a github app to an organization or to a repository. For us to do any validation though that we have the correct installation, we'd need another input variable informing the action of what the app is installed to (or its installation id). Instead I opted to just make this action enforce that the app only have one installation. Another way to say that is if you want to use this action, you should create and install an app specifically for this purpose. Given that constraint, we only need to fetch the first page (which I've also sized to 1 element) since we are only expecting 1 installation.

File: wait.go (hunk 5 above)
Comment (benreynolds-drizly) [reply wait.go]: Updated.
